README.md
agent-base
==========
### Turn a function into an [`http.Agent`][http.Agent] instance
[![Build Status](https://travis-ci.org/TooTallNate/node-agent-base.svg?branch=master)](https://travis-ci.org/TooTallNate/node-agent-base)

This module provides an `http.Agent` generator. That is, you pass it an async
callback function, and it returns a new `http.Agent` instance that will invoke the
given callback function when sending outbound HTTP requests.

#### Some subclasses:

Here's some more interesting uses of `agent-base`.
Send a pull request to list yours!

 * [`http-proxy-agent`][http-proxy-agent]: An HTTP(s) proxy `http.Agent` implementation for HTTP endpoints
 * [`https-proxy-agent`][https-proxy-agent]: An HTTP(s) proxy `http.Agent` implementation for HTTPS endpoints
 * [`pac-proxy-agent`][pac-proxy-agent]: A PAC file proxy `http.Agent` implementation for HTTP and HTTPS
 * [`socks-proxy-agent`][socks-proxy-agent]: A SOCKS (v4a) proxy `http.Agent` implementation for HTTP and HTTPS


Installation
------------

Install with `npm`:

``` bash
$ npm install agent-base
```


Example
-------

Here's a minimal example that creates a new `net.Socket` connection to the server
for every HTTP request (i.e. the equivalent of `agent: false` option):

``` js
var net = require('net');
var tls = require('tls');
var url = require('url');
var http = require('http');
var agent = require('agent-base');

var endpoint = 'http://nodejs.org/api/';
var parsed = url.parse(endpoint);

// This is the important part!
parsed.agent = agent(function (req, opts) {
  var socket;
  // `secureEndpoint` is true when using the https module
  if (opts.secureEndpoint) {
    socket = tls.connect(opts);
  } else {
    socket = net.connect(opts);
  }
  return socket;
});

// Everything else works just like normal...
http.get(parsed, function (res) {
  console.log('"response" event!', res.headers);
  res.pipe(process.stdout);
});
```

API
---

## Agent(Function callback[, Object options]) → [http.Agent][]

Creates a base `http.Agent` that will execute the callback function `callback`
for every HTTP request that it is used as the `agent` for. The callback function
is responsible for creating a `stream.Duplex` instance of some kind that will be
used as the underlying socket in the HTTP request.

The `options` object accepts the following properties:

  * `timeout` - Number - Timeout for the `callback()` function in milliseconds. Defaults to Infinity (optional).

The callback function should have the following signature:

### callback(http.ClientRequest req, Object options, Function cb) → undefined

The ClientRequest `req` can be accessed to read request headers and
and the path, etc. The `options` object contains the options passed
to the `http.request()`/`https.request()` function call, and is formatted
to be directly passed to `net.connect()`/`tls.connect()`, or however
else you want a Socket to be created. Pass the created socket to
the callback function `cb` once created, and the HTTP request will
continue to proceed.

If the `https` module is used to invoke the HTTP request, then the
`secureEndpoint` property on `options` _will be set to `true`_.


License
-------

(The MIT License)

Copyright (c) 2013 Nathan Rajlich &lt;nathan@tootallnate.net&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

[http-proxy-agent]: https://github.com/TooTallNate/node-http-proxy-agent
[https-proxy-agent]: https://github.com/TooTallNate/node-https-proxy-agent
[pac-proxy-agent]: https://github.com/TooTallNate/node-pac-proxy-agent
[socks-proxy-agent]: https://github.com/TooTallNate/node-socks-proxy-agent
[http.Agent]: https://nodejs.org/api/http.html#http_class_http_agent


History.md

4.0.1 / 2017-06-13
==================

  * add `this` context tests and fixes

4.0.0 / 2017-06-06
==================

  * drop support for Node.js < 4
  * drop old versions of Node.js from Travis-CI
  * specify Node.js >= 4.0.0 in `engines.node`
  * remove more old code
  * remove "extend" dependency
  * remove "semver" dependency
  * make the Promise logic a bit cleaner
  * add async function pseudo-example to README
  * use direct return in README example

3.0.0 / 2017-06-02
==================

  * drop support for Node.js v0.8 and v0.10
  * add support for async, Promises, and direct return
  * add a couple `options` test cases
  * implement a `"timeout"` option
  * rename main file to `index.js`
  * test Node 8 on Travis

2.1.1 / 2017-05-30
==================

  * Revert [`fe2162e`](https://github.com/TooTallNate/node-agent-base/commit/fe2162e0ba18123f5b301cba4de1e9dd74e437cd) and [`270bdc9`](https://github.com/TooTallNate/node-agent-base/commit/270bdc92eb8e3bd0444d1e5266e8e9390aeb3095) (fixes #7)

2.1.0 / 2017-05-26
==================

  * unref is not supported for node < 0.9.1 (@pi0)
  * add tests to dangling socket (@pi0)
  * check unref() is supported (@pi0)
  * fix dangling sockets problem (@pi0)
  * add basic "ws" module tests
  * make `Agent` be subclassable
  * turn `addRequest()` into a named function
  * test: Node.js v4 likes to call `cork` on the stream (#3, @tomhughes)
  * travis: test node v4, v5, v6 and v7

2.0.1 / 2015-09-10
==================

  * package: update "semver" to v5.0.1 for WebPack (#1, @vhpoet)

2.0.0 / 2015-07-10
==================

  * refactor to patch Node.js core for more consistent `opts` values
  * ensure that HTTP(s) default port numbers are always given
  * test: use ssl-cert-snakeoil SSL certs
  * test: add tests for arbitrary options
  * README: add API section
  * README: make the Agent HTTP/HTTPS generic in the example
  * README: use SVG for Travis-CI badge

1.0.2 / 2015-06-27
==================

  * agent: set `req._hadError` to true after emitting "error"
  * package: update "mocha" to v2
  * test: add artificial HTTP GET request test
  * test: add artificial data events test
  * test: fix artifical GET response test on node > v0.11.3
  * test: use a real timeout for the async error test

1.0.1 / 2013-09-09
==================

  * Fix passing an "error" object to the callback function on the first tick

1.0.0 / 2013-09-09
==================

  * New API: now you pass a callback function directly

0.0.1 / 2013-07-09
==================

  * Initial release


test/test.js
/**
 * Module dependencies.
 */

var fs = require('fs');
var url = require('url');
var net = require('net');
var tls = require('tls');
var http = require('http');
var https = require('https');
var WebSocket = require('ws');
var assert = require('assert');
var events = require('events');
var inherits = require('util').inherits;
var Agent = require('../');

describe('Agent', function() {
  describe('subclass', function() {
    it('should be subclassable', function(done) {
      function MyAgent() {
        Agent.call(this);
      }
      inherits(MyAgent, Agent);

      MyAgent.prototype.callback = function(req, opts, fn) {
        assert.equal(req.path, '/foo');
        assert.equal(req.getHeader('host'), '127.0.0.1:1234');
        assert.equal(opts.secureEndpoint, true);
        done();
      };

      var info = url.parse('https://127.0.0.1:1234/foo');
      info.agent = new MyAgent();
      https.get(info);
    });
  });
  describe('options', function() {
    it('should support an options Object as first argument', function() {
      var agent = new Agent({ timeout: 1000 });
      assert.equal(1000, agent.timeout);
    });
    it('should support an options Object as second argument', function() {
      var agent = new Agent(function() {}, { timeout: 1000 });
      assert.equal(1000, agent.timeout);
    });
  });
  describe('`this` context', function() {
    it('should be the Agent instance', function(done) {
      var agent = new Agent();
      agent.callback = function () {
        assert.equal(this, agent);
        done();
      }
      var info = url.parse('http://127.0.0.1/foo');
      info.agent = agent;
      http.get(info);
    })
    it('should be the Agent instance with callback signature', function(done) {
      var agent = new Agent();
      agent.callback = function (req, opts, fn) {
        assert.equal(this, agent);
        done();
      }
      var info = url.parse('http://127.0.0.1/foo');
      info.agent = agent;
      http.get(info);
    })
  })
  describe('"error" event', function() {
    it('should be invoked on `http.ClientRequest` instance if `callback()` has not been defined', function(
      done
    ) {
      var agent = new Agent();
      var info = url.parse('http://127.0.0.1/foo');
      info.agent = agent;
      var req = http.get(info);
      req.on('error', function(err) {
        assert.equal(
          '"agent-base" has no default implementation, you must subclass and override `callback()`',
          err.message
        );
        done();
      });
    });
    it('should be invoked on `http.ClientRequest` instance if Error passed to callback function on the first tick', function(
      done
    ) {
      var agent = new Agent(function(req, opts, fn) {
        fn(new Error('is this caught?'));
      });
      var info = url.parse('http://127.0.0.1/foo');
      info.agent = agent;
      var req = http.get(info);
      req.on('error', function(err) {
        assert.equal('is this caught?', err.message);
        done();
      });
    });
    it('should be invoked on `http.ClientRequest` instance if Error passed to callback function after the first tick', function(
      done
    ) {
      var agent = new Agent(function(req, opts, fn) {
        setTimeout(function() {
          fn(new Error('is this caught?'));
        }, 10);
      });
      var info = url.parse('http://127.0.0.1/foo');
      info.agent = agent;
      var req = http.get(info);
      req.on('error', function(err) {
        assert.equal('is this caught?', err.message);
        done();
      });
    });
  });
  describe('artificial "streams"', function() {
    it('should send a GET request', function(done) {
      var stream = new events.EventEmitter();

      // needed for the `http` module to call .write() on the stream
      stream.writable = true;

      stream.write = function(str) {
        assert(0 == str.indexOf('GET / HTTP/1.1'));
        done();
      };

      // needed for `http` module in Node.js 4
      stream.cork = function() {};

      var opts = {
        method: 'GET',
        host: '127.0.0.1',
        path: '/',
        port: 80,
        agent: new Agent(function(req, opts, fn) {
          fn(null, stream);
        })
      };
      var req = http.request(opts);
      req.end();
    });
    it('should receive a GET response', function(done) {
      var stream = new events.EventEmitter();
      var opts = {
        method: 'GET',
        host: '127.0.0.1',
        path: '/',
        port: 80,
        agent: new Agent(function(req, opts, fn) {
          fn(null, stream);
        })
      };
      var req = http.request(opts, function(res) {
        assert.equal('0.9', res.httpVersion);
        assert.equal(111, res.statusCode);
        assert.equal('bar', res.headers.foo);
        done();
      });

      // have to wait for the "socket" event since `http.ClientRequest`
      // doesn't *actually* attach the listeners to the "stream" until
      // this happens
      req.once('socket', function() {
        var buf = new Buffer(
          'HTTP/0.9 111\r\n' +
            'Foo: bar\r\n' +
            'Set-Cookie: 1\r\n' +
            'Set-Cookie: 2\r\n\r\n'
        );
        if ('function' == typeof stream.ondata) {
          // node <= v0.11.3
          stream.ondata(buf, 0, buf.length);
        } else {
          // node > v0.11.3
          stream.emit('data', buf);
        }
      });

      req.end();
    });
  });
});

describe('"http" module', function() {
  var server;
  var port;

  // setup test HTTP server
  before(function(done) {
    server = http.createServer();
    server.listen(0, function() {
      port = server.address().port;
      done();
    });
  });

  // shut down test HTTP server
  after(function(done) {
    server.once('close', function() {
      done();
    });
    server.close();
  });

  it('should work for basic HTTP requests', function(done) {
    var called = false;
    var agent = new Agent(function(req, opts, fn) {
      called = true;
      var socket = net.connect(opts);
      fn(null, socket);
    });

    // add HTTP server "request" listener
    var gotReq = false;
    server.once('request', function(req, res) {
      gotReq = true;
      res.setHeader('X-Foo', 'bar');
      res.setHeader('X-Url', req.url);
      res.end();
    });

    var info = url.parse('http://127.0.0.1:' + port + '/foo');
    info.agent = agent;
    http.get(info, function(res) {
      assert.equal('bar', res.headers['x-foo']);
      assert.equal('/foo', res.headers['x-url']);
      assert(gotReq);
      assert(called);
      done();
    });
  });

  it('should support direct return in `connect()`', function(done) {
    var called = false;
    var agent = new Agent(function(req, opts) {
      called = true;
      return net.connect(opts);
    });

    // add HTTP server "request" listener
    var gotReq = false;
    server.once('request', function(req, res) {
      gotReq = true;
      res.setHeader('X-Foo', 'bar');
      res.setHeader('X-Url', req.url);
      res.end();
    });

    var info = url.parse('http://127.0.0.1:' + port + '/foo');
    info.agent = agent;
    http.get(info, function(res) {
      assert.equal('bar', res.headers['x-foo']);
      assert.equal('/foo', res.headers['x-url']);
      assert(gotReq);
      assert(called);
      done();
    });
  });

  it('should support returning a Promise in `connect()`', function(done) {
    var called = false;
    var agent = new Agent(function(req, opts) {
      return new Promise(function(resolve, reject) {
        called = true;
        resolve(net.connect(opts));
      });
    });

    // add HTTP server "request" listener
    var gotReq = false;
    server.once('request', function(req, res) {
      gotReq = true;
      res.setHeader('X-Foo', 'bar');
      res.setHeader('X-Url', req.url);
      res.end();
    });

    var info = url.parse('http://127.0.0.1:' + port + '/foo');
    info.agent = agent;
    http.get(info, function(res) {
      assert.equal('bar', res.headers['x-foo']);
      assert.equal('/foo', res.headers['x-url']);
      assert(gotReq);
      assert(called);
      done();
    });
  });

  it('should set the `Connection: close` response header', function(done) {
    var called = false;
    var agent = new Agent(function(req, opts, fn) {
      called = true;
      var socket = net.connect(opts);
      fn(null, socket);
    });

    // add HTTP server "request" listener
    var gotReq = false;
    server.once('request', function(req, res) {
      gotReq = true;
      res.setHeader('X-Url', req.url);
      assert.equal('close', req.headers.connection);
      res.end();
    });

    var info = url.parse('http://127.0.0.1:' + port + '/bar');
    info.agent = agent;
    http.get(info, function(res) {
      assert.equal('/bar', res.headers['x-url']);
      assert.equal('close', res.headers.connection);
      assert(gotReq);
      assert(called);
      done();
    });
  });

  it('should pass through options from `http.request()`', function(done) {
    var agent = new Agent(function(req, opts, fn) {
      assert.equal('google.com', opts.host);
      assert.equal('bar', opts.foo);
      done();
    });

    http.get({
      host: 'google.com',
      foo: 'bar',
      agent: agent
    });
  });

  it('should default to port 80', function(done) {
    var agent = new Agent(function(req, opts, fn) {
      assert.equal(80, opts.port);
      done();
    });

    // (probably) not hitting a real HTTP server here,
    // so no need to add a httpServer request listener
    http.get({
      host: '127.0.0.1',
      path: '/foo',
      agent: agent
    });
  });

  it('should support the "timeout" option', function(done) {
    // ensure we timeout after the "error" event had a chance to trigger
    this.timeout(1000);
    this.slow(800);

    var agent = new Agent(
      function(req, opts, fn) {
        // this function will time out
      },
      { timeout: 100 }
    );

    var opts = url.parse('http://nodejs.org');
    opts.agent = agent;

    var req = http.get(opts);
    req.once('error', function(err) {
      assert.equal('ETIMEOUT', err.code);
      req.abort();
      done();
    });
  });
});

describe('"https" module', function() {
  var server;
  var port;

  // setup test HTTPS server
  before(function(done) {
    var options = {
      key: fs.readFileSync(__dirname + '/ssl-cert-snakeoil.key'),
      cert: fs.readFileSync(__dirname + '/ssl-cert-snakeoil.pem')
    };
    server = https.createServer(options);
    server.listen(0, function() {
      port = server.address().port;
      done();
    });
  });

  // shut down test HTTP server
  after(function(done) {
    server.once('close', function() {
      done();
    });
    server.close();
  });

  it('should work for basic HTTPS requests', function(done) {
    var called = false;
    var agent = new Agent(function(req, opts, fn) {
      called = true;
      assert(opts.secureEndpoint);
      var socket = tls.connect(opts);
      fn(null, socket);
    });

    // add HTTPS server "request" listener
    var gotReq = false;
    server.once('request', function(req, res) {
      gotReq = true;
      res.setHeader('X-Foo', 'bar');
      res.setHeader('X-Url', req.url);
      res.end();
    });

    var info = url.parse('https://127.0.0.1:' + port + '/foo');
    info.agent = agent;
    info.rejectUnauthorized = false;
    https.get(info, function(res) {
      assert.equal('bar', res.headers['x-foo']);
      assert.equal('/foo', res.headers['x-url']);
      assert(gotReq);
      assert(called);
      done();
    });
  });

  it('should pass through options from `https.request()`', function(done) {
    var agent = new Agent(function(req, opts, fn) {
      assert.equal('google.com', opts.host);
      assert.equal('bar', opts.foo);
      done();
    });

    https.get({
      host: 'google.com',
      foo: 'bar',
      agent: agent
    });
  });

  it('should default to port 443', function(done) {
    var agent = new Agent(function(req, opts, fn) {
      assert.equal(true, opts.secureEndpoint);
      assert.equal(false, opts.rejectUnauthorized);
      assert.equal(443, opts.port);
      done();
    });

    // (probably) not hitting a real HTTPS server here,
    // so no need to add a httpsServer request listener
    https.get({
      host: '127.0.0.1',
      path: '/foo',
      agent: agent,
      rejectUnauthorized: false
    });
  });
});

describe('"ws" server', function() {
  var wss;
  var server;
  var port;

  // setup test HTTP server
  before(function(done) {
    server = http.createServer();
    wss = new WebSocket.Server({ server: server });
    server.listen(0, function() {
      port = server.address().port;
      done();
    });
  });

  // shut down test HTTP server
  after(function(done) {
    server.once('close', function() {
      done();
    });
    server.close();
  });

  it('should work for basic WebSocket connections', function(done) {
    function onconnection(ws) {
      ws.on('message', function(data) {
        assert.equal('ping', data);
        ws.send('pong');
      });
    }
    wss.on('connection', onconnection);

    var agent = new Agent(function(req, opts, fn) {
      var socket = net.connect(opts);
      fn(null, socket);
    });

    var client = new WebSocket('ws://127.0.0.1:' + port + '/', {
      agent: agent
    });

    client.on('open', function() {
      client.send('ping');
    });

    client.on('message', function(data) {
      assert.equal('pong', data);
      client.close();
      wss.removeListener('connection', onconnection);
      done();
    });
  });
});

describe('"wss" server', function() {
  var wss;
  var server;
  var port;

  // setup test HTTP server
  before(function(done) {
    var options = {
      key: fs.readFileSync(__dirname + '/ssl-cert-snakeoil.key'),
      cert: fs.readFileSync(__dirname + '/ssl-cert-snakeoil.pem')
    };
    server = https.createServer(options);
    wss = new WebSocket.Server({ server: server });
    server.listen(0, function() {
      port = server.address().port;
      done();
    });
  });

  // shut down test HTTP server
  after(function(done) {
    server.once('close', function() {
      done();
    });
    server.close();
  });

  it('should work for secure WebSocket connections', function(done) {
    function onconnection(ws) {
      ws.on('message', function(data) {
        assert.equal('ping', data);
        ws.send('pong');
      });
    }
    wss.on('connection', onconnection);

    var agent = new Agent(function(req, opts, fn) {
      var socket = tls.connect(opts);
      fn(null, socket);
    });

    var client = new WebSocket('wss://127.0.0.1:' + port + '/', {
      agent: agent,
      rejectUnauthorized: false
    });

    client.on('open', function() {
      client.send('ping');
    });

    client.on('message', function(data) {
      assert.equal('pong', data);
      client.close();
      wss.removeListener('connection', onconnection);
      done();
    });
  });
});


index.js
'use strict';

/**
 * Module dependencies.
 */

require('./patch-core');
const inherits = require('util').inherits;
const promisify = require('es6-promisify');
const EventEmitter = require('events').EventEmitter;

/**
 * Module exports.
 */

module.exports = Agent;

/**
 * Base `http.Agent` implementation.
 * No pooling/keep-alive is implemented by default.
 *
 * @param {Function} callback
 * @api public
 */

function Agent(callback, _opts) {
  if (!(this instanceof Agent)) {
    return new Agent(callback, _opts);
  }

  EventEmitter.call(this);

  let opts = _opts;
  if ('function' === typeof callback) {
    this.callback = callback;
  } else if (callback) {
    opts = callback;
  }

  // timeout for the socket to be returned from the callback
  this.timeout = (opts && opts.timeout) || null;
}
inherits(Agent, EventEmitter);

/**
 * Override this function in your subclass!
 */
Agent.prototype.callback = function callback(req, opts) {
  throw new Error(
    '"agent-base" has no default implementation, you must subclass and override `callback()`'
  );
};

/**
 * Called by node-core's "_http_client.js" module when creating
 * a new HTTP request with this Agent instance.
 *
 * @api public
 */

Agent.prototype.addRequest = function addRequest(
  req,
  _opts
) {
  const opts = Object.assign({}, _opts);

  if (opts.host && opts.path) {
    // if both a `host` and `path` are specified then it's most likely the
    // result of a `url.parse()` call... we need to remove the `path` portion so
    // that `net.connect()` doesn't attempt to open that as a unix socket file.
    delete opts.path;
  }

  // set default `port` for HTTP if none was explicitly specified
  if (null == opts.port) {
    opts.port = opts.secureEndpoint ? 443 : 80;
  }

  delete opts.agent;
  delete opts.hostname;
  delete opts._defaultAgent;
  delete opts.defaultPort;
  delete opts.createConnection;

  // hint to use "Connection: close"
  // XXX: non-documented `http` module API :(
  req._last = true;
  req.shouldKeepAlive = false;

  // create the `stream.Duplex` instance
  let timeout;
  let timedOut = false;
  const timeoutMs = this.timeout;

  function onerror(err) {
    if (req._hadError) return;
    req.emit('error', err);
    // For Safety. Some additional errors might fire later on
    // and we need to make sure we don't double-fire the error event.
    req._hadError = true;
  }

  function ontimeout() {
    timedOut = true;
    const err = new Error(
      'A "socket" was not created for HTTP request before ' + timeoutMs + 'ms'
    );
    err.code = 'ETIMEOUT';
    onerror(err);
  }

  function callbackError(err) {
    if (timedOut) return;
    if (timeout != null) {
      clearTimeout(timeout);
    }
    onerror(err);
  }

  function onsocket(socket) {
    if (timedOut) return;
    if (timeout != null) {
      clearTimeout(timeout);
    }
    req.onSocket(socket);
  }

  if (this.callback.length >= 3) {
    // legacy callback function, convert to Promise
    this.callback = promisify(this.callback, this);
  }

  if (timeoutMs > 0) {
    timeout = setTimeout(ontimeout, timeoutMs);
  }

  try {
    Promise.resolve(this.callback(req, opts))
      .then(onsocket, callbackError);
  } catch (err) {
    Promise.reject(err)
      .catch(callbackError);
  }
};


agent.js

/**
 * Module dependencies.
 */

require('./patch-core');
var extend = require('extend');
var inherits = require('util').inherits;
var EventEmitter = require('events').EventEmitter;

/**
 * Module exports.
 */

module.exports = Agent;

/**
 * Base `http.Agent` implementation.
 * No pooling/keep-alive is implemented by default.
 *
 * @param {Function} callback
 * @api public
 */

function Agent (callback) {
  if (!(this instanceof Agent)) return new Agent(callback);
  EventEmitter.call(this);
  if ('function' === typeof callback) {
    this.callback = callback;
  }
}
inherits(Agent, EventEmitter);

Agent.prototype.callback = function callback (req, opts, fn) {
  fn(new Error('"agent-base" has no default implementation, you must subclass and override `callback()`'));
};

/**
 * Called by node-core's "_http_client.js" module when creating
 * a new HTTP request with this Agent instance.
 *
 * @api public
 */

Agent.prototype.addRequest = function addRequest (req, host, port, localAddress) {
  var opts;
  if ('object' == typeof host) {
    // >= v0.11.x API
    opts = extend({}, req._options, host);
  } else {
    // <= v0.10.x API
    opts = extend({}, req._options, { host: host, port: port });
    if (null != localAddress) {
      opts.localAddress = localAddress;
    }
  }

  if (opts.host && opts.path) {
    // if both a `host` and `path` are specified then it's most likely the
    // result of a `url.parse()` call... we need to remove the `path` portion so
    // that `net.connect()` doesn't attempt to open that as a unix socket file.
    delete opts.path;
  }

  // set default `port` if none was explicitly specified
  if (null == opts.port) {
    opts.port = opts.secureEndpoint ? 443 : 80;
  }

  delete opts.agent;
  delete opts.hostname;
  delete opts._defaultAgent;
  delete opts.defaultPort;
  delete opts.createConnection;

  // hint to use "Connection: close"
  // XXX: non-documented `http` module API :(
  req._last = true;
  req.shouldKeepAlive = false;

  // clean up a bit of memory since we're no longer using this
  req._options = null;

  // create the `net.Socket` instance
  var sync = true;
  this.callback(req, opts, function (err, socket) {
    function emitErr () {
      req.emit('error', err);
      // For Safety. Some additional errors might fire later on
      // and we need to make sure we don't double-fire the error event.
      req._hadError = true;
    }
    if (err) {
      if (sync) {
        // need to defer the "error" event, when sync, because by now the `req`
        // instance hasn't event been passed back to the user yet...
        process.nextTick(emitErr);
      } else {
        emitErr();
      }
    } else {
      req.onSocket(socket);
      if (typeof socket.unref === 'function' && process.env.UNREF_SOCKETS) {
        socket.unref();
      }
    }
  });
  sync = false;
};
